File: R/lecture02.R

# STAT 522: Lecture 2 
# Multiple Linear Regression I

library(tidyverse)

# CDA data
# Change varirable names to lower cases
url <- "https://raw.githubusercontent.com/keijioda/Daniel/main/chap10/EXA_C10_S03_01.csv"
cda <- read_csv(url) %>% 
  rename_with(tolower)
cda

# Descriptives
# install.packages("psych")
library(psych)
describe(cda)

# Scatter plot matrix
# install.packages("GGally")
library(GGally)
ggpairs(cda)

# Correlation matrix
cor(cda)

# Run multiple regression
fit <- lm(cda ~ age + edlevel, data = cda)
summary(fit)
anova(fit)

# Using broom package
# install.packages("broom")
library(broom)
tidy(fit)

# Using gtsummary package
# install.packages("gtsummary")
library(gtsummary)
tbl_regression(fit, intercept = TRUE) %>% bold_p()

# Scatterplot with conf interval
# Require visreg package -- install if necessary
# install.packages("visreg")
library(visreg)
par(mfrow = c(2, 1))
  visreg(fit, points=list(cex = 1, pch = 16))
par(mfrow = c(1, 1))

# Matrix approach
intercept <- rep(1, nrow(cda))
x <- as.matrix(cbind(intercept, cda[, 1:2]))
y <- cda$cda
beta <- solve(t(x) %*% x) %*% t(x) %*% y
beta

# Alternatively
solve(crossprod(x, x), crossprod(x, y))

# ANOVA table from regression 
anova(fit)

# R-square and adjusted R-square
summary(fit)
summary(fit)$r.square
summary(fit)$adj.r.square

# Correlation b/w observed and predicted Y
cor(predict(fit), cda$cda)

# 95% CI for each beta coef
confint(fit)

# 95% confidence interval for yhat
predict(fit, interval = 'conf')

# 95% prediction interval for yhat
predict(fit, interval = 'pred')

# 95% Prediction interval for new observations
need_pred <- data.frame(age = c(70, 80), edlevel = c(15, 14))
predict(fit, need_pred, interval = "pred")

File: Python/lecture02.py

# STAT 522: Lecture 2 
# Multiple Linear Regression I

import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np
import scipy.stats as stats

# CDA data
# Change varirable names to lower cases
url = "https://raw.githubusercontent.com/keijioda/Daniel/main/chap10/EXA_C10_S03_01.csv"
cda = pd.read_csv(url)
cda.columns = cda.columns.str.lower()

# Descriptives
cda.describe()

# Scatter plot matrix
sns.pairplot(cda)
plt.show()
plt.clf()

# Correlation matrix
cda.corr()

# Run multiple regression
cda_fit = smf.ols("cda ~ age + edlevel", data = cda).fit()
cda_fit.summary()
sm.stats.anova_lm(cda_fit)

# Scatterplot with conf interval
fig = sm.graphics.plot_partregress_grid(cda_fit, ["age", "edlevel"])
fig.tight_layout(pad = 1.0)
plt.show()

# Matrix approach
x1 = np.ones((cda.shape[0], 1))
x2 = np.array(cda[["age", "edlevel"]])
x = np.concatenate((x1, x2), axis = 1)
y = np.array(cda.cda).reshape(-1, 1)
beta = np.dot(np.linalg.inv(np.dot(x.T, x)), np.dot(x.T, y))
beta

# ANOVA table from regression 
sm.stats.anova_lm(cda_fit)

# R-square and adjusted R-square
# Can be found in the summary table
cda_fit.summary()

# Or can extract them
cda_fit.rsquared
cda_fit.rsquared_adj

# Correlation b/w observed and predicted Y
r, p = stats.pearsonr(cda_fit.predict(), cda.cda)
print('Pearson Corr = %.3f, p = %.3f' % (r, p))

# 95% CI for each beta coef
# Can be found in the summary table
cda_fit.summary()
cda_fit.conf_int()

# 95% confidence interval for yhat
pred = cda_fit.get_prediction(cda[["age", "edlevel"]]).summary_frame()
pred[["mean", "mean_se", "mean_ci_lower", "mean_ci_upper"]]

# 95% prediction interval for yhat
pred[["mean", "mean_se", "obs_ci_lower", "obs_ci_upper"]]

# 95% Prediction interval for new observations
need_pred = pd.DataFrame({"age": [70, 80], "edlevel": [15, 14]})
cda_fit.get_prediction(need_pred).summary_frame()
